File: src/backend/src/controller/project.rs
use crate::{
    dto::{
        AddTeamToProjectRequest, AddTeamToProjectResponse, CreateProjectRequest,
        CreateProjectResponse, DeleteProjectRequest, DeleteProjectResponse, GetProjectRequest,
        GetProjectResponse, ListMyProjectsRequest, ListMyProjectsResponse, ListOrgProjectsRequest,
        ListOrgProjectsResponse, ListProjectTeamsRequest, ListTeamsResponse,
        RemoveTeamFromProjectRequest, RemoveTeamFromProjectResponse, UpdateProjectRequest,
        UpdateProjectResponse,
    },
    service::project_service,
};
use canister_utils::{assert_authenticated, ApiResultDto};
use ic_cdk::{api::msg_caller, *};

#[query]
fn list_my_projects(_: ListMyProjectsRequest) -> ApiResultDto<ListMyProjectsResponse> {
    let caller = msg_caller();
    if let Err(err) = assert_authenticated(&caller) {
        return ApiResultDto::Err(err);
    }

    project_service::list_my_projects(&caller).into()
}

#[query]
fn list_org_projects(req: ListOrgProjectsRequest) -> ApiResultDto<ListOrgProjectsResponse> {
    let caller = msg_caller();
    if let Err(err) = assert_authenticated(&caller) {
        return ApiResultDto::Err(err);
    }

    project_service::list_org_projects(&caller, req).into()
}

#[update]
fn create_project(req: CreateProjectRequest) -> ApiResultDto<CreateProjectResponse> {
    let caller = msg_caller();
    if let Err(err) = assert_authenticated(&caller) {
        return ApiResultDto::Err(err);
    }

    project_service::create_project(&caller, req).into()
}

#[query]
fn get_project(req: GetProjectRequest) -> ApiResultDto<GetProjectResponse> {
    let caller = msg_caller();
    if let Err(err) = assert_authenticated(&caller) {
        return ApiResultDto::Err(err);
    }

    project_service::get_project(&caller, req).into()
}

#[update]
fn update_project(req: UpdateProjectRequest) -> ApiResultDto<UpdateProjectResponse> {
    let caller = msg_caller();
    if let Err(err) = assert_authenticated(&caller) {
        return ApiResultDto::Err(err);
    }

    project_service::update_project(&caller, req).into()
}

#[update]
fn delete_project(req: DeleteProjectRequest) -> ApiResultDto<DeleteProjectResponse> {
    let caller = msg_caller();
    if let Err(err) = assert_authenticated(&caller) {
        return ApiResultDto::Err(err);
    }

    project_service::delete_project(&caller, req).into()
}

#[query]
fn list_project_teams(req: ListProjectTeamsRequest) -> ApiResultDto<ListTeamsResponse> {
    let caller = msg_caller();
    if let Err(err) = assert_authenticated(&caller) {
        return ApiResultDto::Err(err);
    }

    project_service::list_project_teams(&caller, req).into()
}

#[update]
fn add_team_to_project(req: AddTeamToProjectRequest) -> ApiResultDto<AddTeamToProjectResponse> {
    let caller = msg_caller();
    if let Err(err) = assert_authenticated(&caller) {
        return ApiResultDto::Err(err);
    }

    project_service::add_team_to_project(&caller, req).into()
}

#[update]
fn remove_team_from_project(
    req: RemoveTeamFromProjectRequest,
) -> ApiResultDto<RemoveTeamFromProjectResponse> {
    let caller = msg_caller();
    if let Err(err) = assert_authenticated(&caller) {
        return ApiResultDto::Err(err);
    }

    project_service::remove_team_from_project(&caller, req).into()
}

File: src/frontend/src/lib/api/project.ts
import {
  mapAddTeamToProjectRequest,
  mapAddTeamToProjectResponse,
  mapCreateProjectRequest,
  mapCreateProjectResponse,
  mapDeleteProjectRequest,
  mapDeleteProjectResponse,
  mapGetProjectRequest,
  mapGetProjectResponse,
  mapListMyProjectsResponse,
  mapListOrgProjectsRequest,
  mapListOrgProjectsResponse,
  mapListProjectTeamsRequest,
  mapListProjectTeamsResponse,
  mapRemoveTeamFromProjectRequest,
  mapRemoveTeamFromProjectResponse,
  mapUpdateProjectRequest,
  mapUpdateProjectResponse,
  mapUpdateTeamProjectPermissionsRequest,
  mapUpdateTeamProjectPermissionsResponse,
  type AddTeamToProjectRequest,
  type CreateProjectRequest,
  type DeleteProjectRequest,
  type GetProjectRequest,
  type ListMyProjectsResponse,
  type ListOrgProjectsRequest,
  type ListOrgProjectsResponse,
  type ListProjectTeamsRequest,
  type ListProjectTeamsResponse,
  type ProjectResponse,
  type ProjectTeamResponse,
  type RemoveTeamFromProjectRequest,
  type UpdateProjectRequest,
  type UpdateTeamProjectPermissionsRequest,
} from '@/lib/api-models';
import type { ActorSubclass } from '@icp-sdk/core/agent';
import type { _SERVICE } from '@ssn/backend-api';

export class ProjectApi {
  constructor(private readonly actor: ActorSubclass<_SERVICE>) {}

  public async listMyProjects(): Promise<ListMyProjectsResponse> {
    const res = await this.actor.list_my_projects({});
    return mapListMyProjectsResponse(res);
  }

  public async listOrgProjects(
    req: ListOrgProjectsRequest,
  ): Promise<ListOrgProjectsResponse> {
    const res = await this.actor.list_org_projects(
      mapListOrgProjectsRequest(req),
    );
    return mapListOrgProjectsResponse(res);
  }

  public async createProject(
    req: CreateProjectRequest,
  ): Promise<ProjectResponse> {
    const res = await this.actor.create_project(mapCreateProjectRequest(req));
    return mapCreateProjectResponse(res);
  }

  public async getProject(req: GetProjectRequest): Promise<ProjectResponse> {
    const res = await this.actor.get_project(mapGetProjectRequest(req));
    return mapGetProjectResponse(res);
  }

  public async updateProject(
    req: UpdateProjectRequest,
  ): Promise<ProjectResponse> {
    const res = await this.actor.update_project(mapUpdateProjectRequest(req));
    return mapUpdateProjectResponse(res);
  }

  public async deleteProject(req: DeleteProjectRequest): Promise<void> {
    const res = await this.actor.delete_project(mapDeleteProjectRequest(req));
    mapDeleteProjectResponse(res);
  }

  public async listProjectTeams(
    req: ListProjectTeamsRequest,
  ): Promise<ListProjectTeamsResponse> {
    const res = await this.actor.list_project_teams(
      mapListProjectTeamsRequest(req),
    );
    return mapListProjectTeamsResponse(res);
  }

  public async addTeamToProject(req: AddTeamToProjectRequest): Promise<void> {
    const res = await this.actor.add_team_to_project(
      mapAddTeamToProjectRequest(req),
    );
    mapAddTeamToProjectResponse(res);
  }

  public async removeTeamFromProject(
    req: RemoveTeamFromProjectRequest,
  ): Promise<void> {
    const res = await this.actor.remove_team_from_project(
      mapRemoveTeamFromProjectRequest(req),
    );
    mapRemoveTeamFromProjectResponse(res);
  }

  public async updateTeamProjectPermissions(
    req: UpdateTeamProjectPermissionsRequest,
  ): Promise<ProjectTeamResponse> {
    const res = await this.actor.update_team_project_permissions(
      mapUpdateTeamProjectPermissionsRequest(req),
    );
    return mapUpdateTeamProjectPermissionsResponse(res);
  }
}
